[gps] Add hdop sensor

realzoulou opened the pull request:
# What does this implement/fix?
TinyGPSPlus in currently used version 1.0.2 supports reading HDOP (Horizontal Dilution Of Precision).
However the current `gps` component does not define a sensor for it.
This add `hdop` as an additional sensor for `gps`.
Additional changes
- Change debug logs to 1 line per value instead of 2
- Change defaults from -1 to NAN, satellites defaults to 0

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes https://github.com/esphome/feature-requests/issues/3140

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4882

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
  hdop:
    name: "HDOP"
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in https://github.com/esphome/esphome-docs/pull/4879

## Tested with
1. Heltec HTIT Tracker (https://heltec.org/project/wireless-tracker) and its onboard UC6580 GNSS chipset
2. Same Heltec, but with external u-blox NEO-F10N (https://www.sparkfun.com/sparkfun-gnss-l1-l5-breakout-neo-f10n-sma.html)

realzoulou committed:
Add hdop sensor (#3140)

Additional changes
- Change debug logs to 1 line per value instead of 2
- Change defaults from -1 to NAN, satellites defaults to 0

realzoulou committed:
Add hdop sensor (#3140)

Additional changes
- Change  some debug logs to 1 line per value instead of 2
- Change defaults from -1 to NAN, satellites defaults to 0

coogle commented:
Sure I can take a peek.

coogle commented on esphome/components/gps/__init__.py:
```diff
@@ -40,20 +41,24 @@
             ),
             cv.Optional(CONF_SPEED): sensor.sensor_schema(
                 unit_of_measurement=UNIT_KILOMETER_PER_HOUR,
-                accuracy_decimals=6,
```
Why have you changed the accuracy here?

realzoulou replied:
Speed accuracy in GNSS is far below 6 decimals. Even 2 decimals are too much accuracy. If you disagree then I will revert back to original 6 decimals.

coogle replied:
Nah, if there is a good reason for it I'm fine with it. However, according to [this](https://www.gps.gov/systems/gps/performance/accuracy)

> The government provides the GPS signal in space with a global average user range rate error (URRE) of ≤0.006 m/sec over any 3-second interval, with 95% probability.

So we should probably use that as the precision?

realzoulou replied:
I updated to 3 decimals

coogle commented on esphome/components/gps/gps.cpp:
```diff
@@ -37,30 +40,29 @@ void GPS::loop() {
 
-        ESP_LOGD(TAG, "Location:");
-        ESP_LOGD(TAG, "  Lat: %f", this->latitude_);
-        ESP_LOGD(TAG, "  Lon: %f", this->longitude_);
```
No need to change the spacing on these, please leave as-is

coogle reviewed (changes requested):
Overall looks fine, but I think you've made some minor changes that were unnecessary and I'd like to see those undone unless there is a reason for them.

realzoulou committed:
Add hdop sensor (#3140)

Additional changes
- Change  some debug logs to 1 line per value instead of 2
- Change defaults from -1 to NAN, satellites defaults to 0

jesserockz commented on esphome/components/gps/__init__.py:
```diff
@@ -92,5 +97,9 @@ async def to_code(config):
         sens = await sensor.new_sensor(config[CONF_SATELLITES])
         cg.add(var.set_satellites_sensor(sens))
 
+    if CONF_HDOP in config:
+        sens = await sensor.new_sensor(config[CONF_HDOP])
```
```suggestion
    if hdop_config := config.get(CONF_HDOP):
        sens = await sensor.new_sensor(hdop_config)
```

jesserockz committed:
Update esphome/components/gps/__init__.py

jesserockz commented:
Thanks for looking over @coogle
